server/utils/csvDownloaderUtil.js

```diff
@@ -0,0 +1,23 @@
+import { stringify } from 'csv-stringify'
+
+export default async function csvDownloader(res, { tableColumns, tableRows }) {
```

Comment:
instead of continuing to throw functions into utils, I think we should look for opportunities to create more specific types of objects. For example, I wonder if this would make sense as a service object. I'm thinking something along the lines of

```js
// server/services/CsvService.js

class CsvService {
  constructor(csvData) {
    const { tableColumns, tableRows } = csvData
    if (!tableColumns || !tableRows) {
      throw new Error('CSV data must have tableColumns and tableRows attributes')
    }

    this.columns = tableColumns
    this.rows = tableRows
  }

  get csvDataString() {
    return stringify(this.rows, {
      header: true,
      columns: this.columns,
    })
  }

  toCsv = async () => {
    return new Promise((resolve, reject) => {
      const data = this.csvDataString
      const rows = []

      data.on('readable', () => {
        let row = tableDataToCsvString.read()
        while (row !== null) {
          rows.push(row)
          row = tableDataToCsvString.read()
        }
      })

      data.on('finish', () => resolve(rows))
      data.on('error', (err) => reject(err))
    })
  }
}

export default CsvService
```

Then in the controller you could use the service to generate the CSV

```js
const parsedQueryParams = qs.parse(req.query)
const { graphTable } = await graphDataController(
  dataDb,
  userAccess,
  chart_id,
  parsedQueryParams
)
const csvService = new CsvService(graphTable)
const csvData = await csvService.toCsv()
res.write(csvData)
res.send()
```

what do you think about this structure?

Reply:
Thanks for this @mikestone14, I am not opposed to it instead I am curious. In the coming days could we just have a chat on this? What are some of the use cases we would consider this approach over writing functions. 
Changes are applied.
https://github.com/AMP-SCZ/dpdash/pull/287/commits/53c75c19215f81ea1f646fb5e98102019a21521f

Reply:
https://github.com/AMP-SCZ/dpdash/commit/1495cca43f24bbd29b4c287f1e1fb00d1d16b938